### PR TITLE
update docker's resolv.conf file with options ndots:5

### PR DESCRIPTION
**Pull request by ArtfulCoder**
Proof-of-concept PR for #10161 issue.
superseded #10241 PR
- manually verified that it works
- PR needs to be fixed up since ndots entry is done twice in the /etc/resolv.conf file(innocuous) amongst other things
- the earlier PR was more complex where we would unmount docker's resolv.conf file and mount our own
- in this PR, we simply modify the /etc/resolv.conf file that docker is mounting.

### CODE REVIEW

**Comment by ArtfulCoder** [comment]
@thockin @dchen1107 @smarterclayton

**Comment by ArtfulCoder** [comment]
In this PR, we simply add the "options ndots:3" to the docker generated resolv.conf file.
Far simpler that the #10241 PR which requires mount/umount

**Comment by smarterclayton** [comment]
The challenge is now that user code can race against the resolver - if someone has a very fast script that makes a curl call, we probably won't beat them.  This makes it *possible* to use ndots, so it's an improvement, but I worry that it will be unreliable for users.

**Comment by ArtfulCoder** [comment]
In reality, all containers of the pod share the same dns resolv file, because docker sees that the other containers have marked the pause container as the container to use for networking namespace.

The pause container gets created first and we update the resolv file for the pause container.
Other containers of the pod get created later with the updated resolv file.

So we should not have this problem.

**Comment by smarterclayton** [comment]
Is there no case today where we recreate the infra pod without tearing down the old containers?

**Comment by ArtfulCoder** [comment]
If the infra pod is teared down, the networking namespace goes with it and old containers of the same pod have to be recreated.

File: pkg/kubelet/dockertools/manager.go [hunk 1]
```diff
@@ -28,6 +28,7 @@ import (
 	"strconv"
 	"strings"
 	"sync"
+	//"syscall"
```
Comment (smarterclayton) [inline pkg/kubelet/dockertools/manager.go]: Remove this

File: pkg/kubelet/dockertools/manager.go [hunk 2]
```diff
@@ -1223,10 +1224,61 @@ func (dm *DockerManager) runContainerInPod(pod *api.Pod, container *api.Containe
...
+	ndotsOption := "options ndots:3\n"
+	for i := 0; i < 10; i++ {
+		err := appendToFile(resolvFilePath, ndotsOption)
+		if err == nil {
+			glog.Infof("resolv.conf updated with ndots option for container:%s", containerName)
+			return nil
+		}
+		time.Sleep(time.Second)
+	}
+	glog.Errorf("resolv.conf could not be updated for container:%s", containerName)
```
Comment (smarterclayton) [inline pkg/kubelet/dockertools/manager.go]: Nit: space between : and %s

File: pkg/kubelet/dockertools/manager.go [hunk 3]
```diff
@@ -1223,10 +1224,61 @@ func (dm *DockerManager) runContainerInPod(pod *api.Pod, container *api.Containe
...
+	}
+
+	if !fileExists(resolvFilePath) {
+		glog.Errorf("DNS ResolvConfPath does not exist and could not be updated. Path:%s, Container:%s. ", resolvFilePath, containerName)
+		return nil
+	}
+
+	glog.Infof("DNS ResolvConfPath exists:%s for container:%s. Will attempt to add ndots:3 option.", resolvFilePath, containerName)
+
+	ndotsOption := "options ndots:3\n"
```
Comment (smarterclayton) [inline pkg/kubelet/dockertools/manager.go]: Make this a package private constant

**Comment by ArtfulCoder** [comment]
note to self: tests, address comments, add a lock to not update the same file twice.

File: pkg/kubelet/dockertools/manager.go [hunk 4]
```diff
@@ -1223,10 +1224,61 @@ func (dm *DockerManager) runContainerInPod(pod *api.Pod, container *api.Containe
...
+	return nil
+}
+
+func fileExists(fileName string) bool {
+	_, err := os.Stat(fileName)
+	return os.IsExist(err)
+}
+
+func appendToFile(filePath, stringToAppend string) error {
+	f, err := os.OpenFile(filePath, os.O_APPEND|os.O_WRONLY, 0644)
```
Comment (smarterclayton) [inline pkg/kubelet/dockertools/manager.go]: Will this fail if the file doesn't exist (due to another issue, like docker getting stopped and tearing down the container) and if so will it be a clean error?

File: pkg/kubelet/dockertools/manager.go [hunk 5]
```diff
@@ -1223,10 +1224,61 @@ func (dm *DockerManager) runContainerInPod(pod *api.Pod, container *api.Containe
...
+	}
+	defer f.Close()
+
+	//err = syscall.Flock(f.Fd(), LOCK_EX)
+
+	_, err = f.WriteString(stringToAppend)
+	if err != nil {
+		glog.Errorf("Error writing to file %s. err: %v", filePath, err)
+	} else {
+		glog.Infof("Appended text:'%s' to file %s ", stringToAppend, filePath)
```
Comment (smarterclayton) [inline pkg/kubelet/dockertools/manager.go]: V(4) or don't print at all

File: pkg/kubelet/dockertools/manager.go [hunk 6]
```diff
@@ -1223,10 +1224,61 @@ func (dm *DockerManager) runContainerInPod(pod *api.Pod, container *api.Containe
...
+		glog.Errorf("DNS ResolvConfPath does not exist and could not be updated. Path:%s, Container:%s. ", resolvFilePath, containerName)
+		return nil
+	}
+
+	glog.Infof("DNS ResolvConfPath exists:%s for container:%s. Will attempt to add ndots:3 option.", resolvFilePath, containerName)
+
+	ndotsOption := "options ndots:3\n"
+	for i := 0; i < 10; i++ {
+		err := appendToFile(resolvFilePath, ndotsOption)
+		if err == nil {
```
Comment (smarterclayton) [inline pkg/kubelet/dockertools/manager.go]: Nit: style would recommend 

    if err := appendToFile(...); err == nil

Instead

File: pkg/kubelet/dockertools/manager.go [hunk 7]
```diff
@@ -1223,10 +1224,61 @@ func (dm *DockerManager) runContainerInPod(pod *api.Pod, container *api.Containe
...
+		glog.Errorf("Error opening file %s for appending. err: %v", filePath, err)
+		return err
+	}
+	defer f.Close()
+
+	//err = syscall.Flock(f.Fd(), LOCK_EX)
+
+	_, err = f.WriteString(stringToAppend)
+	if err != nil {
+		glog.Errorf("Error writing to file %s. err: %v", filePath, err)
```
Comment (smarterclayton) [inline pkg/kubelet/dockertools/manager.go]: I don't think we should log this except on our last loop - the caller should log it

**Comment by smarterclayton** [comment]
What would prevent us from updating the file?  Kubelet killed during restart?  I would expect only one execution path here.

**Comment by thockin** [comment]
I haven't done review yet, but pathologically, we need to set ndots to 5 if
we want `_dns._udp.kube-dns.default.svc` to resolve.

**Comment by ArtfulCoder** [comment]
@k8s-bot test this please

File: pkg/kubelet/dockertools/manager.go [hunk 8]
```diff
@@ -59,6 +59,7 @@ const (
 
 	kubernetesPodLabel       = "io.kubernetes.pod.data"
 	kubernetesContainerLabel = "io.kubernetes.container.name"
+	ndotsDNSOption           = "options ndots:5\n"
```
Comment (thockin) [inline pkg/kubelet/dockertools/manager.go]: comment why 5 is the right value

**Comment by thockin** [comment]
This is a Good Hack.  It's clever, but actually pretty safe.  Please add some good comments on addNDotsOption() that explains what we're doing and why and what we need to be able to stop doing this.

And yeah, needs a test so we don't rely on something that is really an implementation detail and might go away in the future.

**Comment by ArtfulCoder** [comment]
Added e2e test.
The test did not pass in my test cluster. Checking to see if k8s-bot gives a better result :)

**Comment by ArtfulCoder** [comment]
there is a bug in the test code changes (not the src code). Fixing test.

**Comment by ArtfulCoder** [comment]
@thockin @smarterclayton Thanks for the PR comments. I think have addressed them all.
The e2e test is updated and should be green in the next run. (It is green on my test cluster)
I did a small refactor as well in the e2e test just because I was in there..

**Comment by thockin** [comment]
LGTM.  We can reduce the test case back to one when we do a new dnsutils container.

**Comment by thockin** [comment]
@smarterclayton if you're not around, I'll get @dchen1107 to provide 2nd LGTM

**Comment by ArtfulCoder** [comment]
@thockin I already created a new dnsutils container. It is part of the PR. Its called jessie-dnsutils..

I didn't quite follow your new-container-comment..

**Comment by ArtfulCoder** [comment]
@k8s-bot test this please

**Comment by thockin** [comment]
I want to build a new dnsutils container that is not based on debian at
all.  when I do that this test is not that useful any more.

**Comment by dchen1107** [comment]
ok to test

**Comment by ArtfulCoder** [comment]
@thockin ok, thanks

**Comment by dchen1107** [comment]
LGTM

**Comment by ArtfulCoder** [comment]
@k8s-bot test this please
ok to test

**Comment by ArtfulCoder** [comment]
http://kubekins.dls.corp.google.com/view/Critical%20Builds/job/kubernetes-e2e-gce/7202/

Jenkins GCE e2e passed.

**Comment by ghaskins** [comment]
This change appears to have broken kubernetes in my environment w.r.t. accessing external DNS for reasons I do not fully understand yet.  Is there a way to make this optional?

**Comment by thockin** [comment]
@ghaskins only if you don't want kubernetes DNS to work. You can probably reduce it from 5 to 3, but that may change again in the future...

**Comment by karteek** [comment]
This change seems to make external DNS lookups very slow.
Any external DNS request with less than 5 dots, seem to make `n` requests (number of search domains) before resolving correctly.

```
root@mypod-controller-203fa# nslookup -debug
> google.com
Server:     10.3.0.10
Address:    10.3.0.10#53

------------
    QUESTIONS:
    google.com.my-namespace.svc.cluster.local, type = A, class = IN
    ANSWERS:
    AUTHORITY RECORDS:
    ->  cluster.local
    origin = ns.dns.cluster.local
    mail addr = hostmaster.cluster.local
    serial = 1470690000
    refresh = 28800
    retry = 7200
    expire = 604800
    minimum = 60
    ttl = 60
    ADDITIONAL RECORDS:
------------
** server can't find google.com.my-namespace.svc.cluster.local: NXDOMAIN
Server:     10.3.0.10
Address:    10.3.0.10#53

------------
    QUESTIONS:
    google.com.svc.cluster.local, type = A, class = IN
    ANSWERS:
    AUTHORITY RECORDS:
    ->  cluster.local
    origin = ns.dns.cluster.local
    mail addr = hostmaster.cluster.local
    serial = 1470690000
    refresh = 28800
    retry = 7200
    expire = 604800
    minimum = 60
    ttl = 60
    ADDITIONAL RECORDS:
------------
** server can't find google.com.svc.cluster.local: NXDOMAIN
Server:     10.3.0.10
Address:    10.3.0.10#53

------------
    QUESTIONS:
    google.com.cluster.local, type = A, class = IN
    ANSWERS:
    AUTHORITY RECORDS:
    ->  cluster.local
    origin = ns.dns.cluster.local
    mail addr = hostmaster.cluster.local
    serial = 1470690000
    refresh = 28800
    retry = 7200
    expire = 604800
    minimum = 60
    ttl = 60
    ADDITIONAL RECORDS:
------------
** server can't find google.com.cluster.local: NXDOMAIN
Server:     10.3.0.10
Address:    10.3.0.10#53

------------
    QUESTIONS:
    google.com.us-west-2.compute.internal, type = A, class = IN
    ANSWERS:
    AUTHORITY RECORDS:
    ADDITIONAL RECORDS:
------------
** server can't find google.com.us-west-2.compute.internal: NXDOMAIN
Server:     10.3.0.10
Address:    10.3.0.10#53

------------
    QUESTIONS:
    google.com, type = A, class = IN
    ANSWERS:
    ->  google.com
    internet address = 172.217.3.174
    ttl = 173
    AUTHORITY RECORDS:
    ADDITIONAL RECORDS:
------------
Non-authoritative answer:
Name:   google.com
Address: 172.217.3.174
```

Is there any way, I can use kubernetes DNS without slowing down external DNS resolution?

**Comment by thockin** [comment]
Can you characterize "slow"  I see low milliseconds for DNS lookups, even
with ndots 5

**Comment by karteek** [comment]
I don't have exact numbers to quantify the slowness.
But, on high scale, each request to an external REST api gets 20-100 milliseconds slower which adds up.

This is primarily due to two reasons IMO
1. One is due to ndots 5,
2. KubeDNS doesn't seem to be caching DNS results (https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-dns/app/server.go#L140)

**Comment by thockin** [comment]
kube-dns has a dnsmasq in front of it doign caching, supposedly.

**Comment by girishkalele** [comment]
@karteek 

There is dnsmasq caching in 1.3 - could you run a describe on your kube-dns RC ?

There are two questions in my mind that I need to verify -
 - latency of a negative response from kube-dns should be extremely low
 - does dnsmasq cache negative responses from kube-dns to speed up the progression through the N searches

**Comment by karteek** [comment]
@girishkalele 

I understand that there is dnsmasq caching. Probably thats why I didn't see much difference in dnsperf results even after enabling caching on kube-dns.

This is my kube-dns RC

```
➜  ~ kubectl --namespace=kube-system describe rc kube-dns-v19
Name:		kube-dns-v19
Namespace:	kube-system
Image(s):	gcr.io/google_containers/kubedns-amd64:1.6,gcr.io/google_containers/kube-dnsmasq-amd64:1.3,gcr.io/google_containers/exechealthz-amd64:1.1
Selector:	k8s-app=kube-dns,version=v19
Labels:		k8s-app=kube-dns
		kubernetes.io/cluster-service=true
		version=v19
Replicas:	12 current / 12 desired
Pods Status:	12 Running / 0 Waiting / 0 Succeeded / 0 Failed
No volumes.
No events.
```
I did compile kube-dns on my macbook pro, and ran it as `./kube-dns --dns-port=10053 --kubecfg-file=kubeconfig` and ran dns-perf against it.  

```
Statistics:

  Queries sent:         2616456
  Queries completed:    2616456 (100.00%)
  Queries lost:         0 (0.00%)

  Response codes:       NOERROR 2616456 (100.00%)
  Average packet size:  request 46, response 100
  Run time (s):         60.585622
  Queries per second:   43186.087947

  Average Latency (s):  0.002282 (min 0.008085, max 53.097277)
  Latency StdDev (s):   0.003720
```

Same against kube-dns service in a kubernetes cluster running on AWS

```
Statistics:

  Queries sent:         145273
  Queries completed:    145241 (99.98%)
  Queries lost:         32 (0.02%)

  Response codes:       NOERROR 145241 (100.00%)
  Average packet size:  request 46, response 103
  Run time (s):         60.253472
  Queries per second:   2410.500095

  Average Latency (s):  0.040293 (min 3.253098, max 33.360985)
  Latency StdDev (s):   0.016708
```

**Pull request merged**